cmd/activator/main.go

```diff
@@ -133,6 +150,70 @@ func main() {
 		logger.Fatalw("Failed to create stats reporter", zap.Error(err))
 	}
 
+	// Set up signals so we handle the first shutdown signal gracefully.
+	stopCh := signals.SetupSignalHandler()
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, defaultResyncInterval)
+	servingInformerFactory := servinginformers.NewSharedInformerFactory(servingClient, defaultResyncInterval)
+	endpointInformer := kubeInformerFactory.Core().V1().Endpoints()
+	revisionInformer := servingInformerFactory.Serving().V1alpha1().Revisions()
+
+	go endpointInformer.Informer().Run(stopCh)
+	go revisionInformer.Informer().Run(stopCh)
+
+	params := queue.BreakerParams{QueueDepth: breakerQueueDepth, MaxConcurrency: breakerMaxConcurrency, InitialCapacity: 0}
+
+	// Return the number of endpoints, 0 if no endpoints are found.
+	endpointsGetter := func(revID activator.RevisionID) (int32, error) {
+		endpoints, err := endpointInformer.Lister().Endpoints(revID.Namespace).Get(revID.Name)
+		if errors.IsNotFound(err) {
+			return 0, nil
+		}
+		if err != nil {
+			return 0, err
+		}
+		addresses := activator.EndpointsAddressCount(endpoints.Subsets)
+		return int32(addresses), nil
+	}
+
+	// Return the revision from the observer.
+	revisionGetter := func(revID activator.RevisionID) (*v1alpha12.Revision, error) {
+		rev, err := revisionInformer.Lister().Revisions(revID.Namespace).Get(revID.Name)
+		if err != nil {
+			return nil, err
+		}
+		return rev, err
+	}
+
+	throttlerParams := activator.ThrottlerParams{BreakerParams: params, Logger: logger, GetEndpoints: endpointsGetter, GetRevision: revisionGetter}
+	throttler := activator.NewThrottler(throttlerParams)
+
+	// Update/create the breaker in the throttler when the number of endpoints changes.
+	endpointInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    activator.UpdateEndpoints(throttler),
+		UpdateFunc: controller.PassNew(activator.UpdateEndpoints(throttler)),
+	})
+
+	// Remove the breaker if the revision was deleted.
+	revisionInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: activator.DeleteBreaker(throttler),
```

Comment:
Why not simply do this on Endpoints as well?

Reply:
I'm not sure whether we need to do that, if an endpoint is deleted it could simply mean that we scaled down to 0 and we would need the breaker once new requests come in.

Reply:
I believe the Endpoints resource is deleted iff the K8s Service is deleted, and we don't do that unless the Revision is deleted.

Reply:
Got it, will update it to delete the breaker on the Endpoints.